lib/matplotlib/colorbar.py

```diff
@@ -510,166 +490,286 @@ def __init__(self, ax, cmap=None,
+        self.ax.add_collection(col)
         self.stale = True
 
-    config_axis = _api.deprecate_privatize_attribute("3.3")
+    def update_ticks(self):
+        """
+        Setup the ticks and ticklabels. This should not be needed by users.
+        """
+        ax = self.ax
+        # Get the locator and formatter; defaults to self.locator if not None.
+        self._get_ticker_locator_formatter()
+        self._long_axis().set_major_locator(self.locator)
+        self._long_axis().set_minor_locator(self.minorlocator)
+        self._long_axis().set_major_formatter(self.formatter)
 
     def _get_ticker_locator_formatter(self):
         """
         Return the ``locator`` and ``formatter`` of the colorbar.
 
-        If they have not been defined (i.e. are *None*), suitable formatter
-        and locator instances will be created, attached to the respective
-        attributes and returned.
+        If they have not been defined (i.e. are *None*), the formatter and
+        locator are retrieved from the axis, or from the value of the
+        boundaries for a boundary norm.
+
+        Called by update_ticks...
         """
         locator = self.locator
         formatter = self.formatter
-        if locator is None:
-            if self.boundaries is None:
-                if isinstance(self.norm, colors.NoNorm):
-                    nv = len(self._values)
-                    base = 1 + int(nv / 10)
-                    locator = ticker.IndexLocator(base=base, offset=0)
-                elif isinstance(self.norm, colors.BoundaryNorm):
-                    b = self.norm.boundaries
-                    locator = ticker.FixedLocator(b, nbins=10)
-                elif isinstance(self.norm, colors.LogNorm):
-                    locator = _ColorbarLogLocator(self)
-                elif isinstance(self.norm, colors.SymLogNorm):
-                    # The subs setting here should be replaced
-                    # by logic in the locator.
-                    locator = ticker.SymmetricalLogLocator(
-                                      subs=np.arange(1, 10),
-                                      linthresh=self.norm.linthresh,
-                                      base=10)
-                else:
-                    if mpl.rcParams['_internal.classic_mode']:
-                        locator = ticker.MaxNLocator()
-                    else:
-                        locator = _ColorbarAutoLocator(self)
-            else:
-                b = self._boundaries[self._inside]
+        minorlocator = self.minorlocator
+        if isinstance(self.norm, colors.BoundaryNorm):
+            b = self.norm.boundaries
+            if locator is None:
                 locator = ticker.FixedLocator(b, nbins=10)
+        elif self.boundaries is not None:
+            b = self._boundaries[self._inside]
+            if locator is None:
+                locator = ticker.FixedLocator(b, nbins=10)
+        else:  # most cases:
```

Comment:
This looks circular: `update_ticks` calls *this* function to find the locator, etc, *before* setting them on the long axis; but this function *looks* for them on the long axis.  I suspect the circular case doesn't occur in practice, but it looks odd when reading the code.  Maybe add a comment?

Reply:
Sure, updated to 
```
            if locator is None:
                # we haven't set the locator explicitly, so use the default 
                # for this axis:
```

This is really just a dumb way to get the default locator for the axis....